### DISCUSSION TITLE
TM129 container discussion

### DISCUSSION THREAD

**Discussion by mmh352** [mmh352/ou-container-builder#28] [General]
Okay, so I need to clarify (from #19):

1. the container I built for 21B for deployment by you won't work as is from 21J? This is really important to know, because it would require a sweep of all the materials looking for jovyan references, or move to the SSU server.

2. it seems that there is an assumption of one JupyterHub to rule them all, rather than one JupyterHub per module/course, which is what I think some institutions do? (Eg from the existence of https://github.com/yuvipanda/hubploy, I wonder if Berkeley run separate JupyterHubs? But I guess that then causes issues about how you create a single volume for a student that you can mount to different JupyterHubs for the different courses they are on?) I'm not sure what the SSU or IT take is on this or whether due consideration has been made about the pros and cons of each approach (eg taking into account that modules are isolated and have different presentation patterns, that students defer on a module or may be on several modules at once, that students may want unified storage or be happy with per module storage, etc). Given the OU module is still dominated by module team fiefdoms, the one-jupyter-hub-per-module (or even per-module-per-presentation) seemed to for that structure and potentially gave module teams ownership of their environment isolated from everyone else.

3. by "the JupyterHub on k8s installation", I assume you mean "the mmh352 JupyterHub on k8s installation", rather than the "the SSU JupyterHub on k8s installation" or "the IT JupyterHub on k8s installation", or "an arbitrary zero-to-JupyterHub on k8s installation"? I'm trying to understand what gets set where and why. And whether the setting is fixed or could be set via a parameter (and how/where that parameter would get set in turn). I imagine a lot of set up is around the UID, which we can take out of the equation as long as we agree on UID. So:
    * is the username required just to set the mount path, or for other reasons / in other places as well? I get that currently the user is baked in at image build time, but I don't see a reason why you couldn't in principle create a user with a customised name on a container's first run, with the name set from an ENV variable set via the container invocation, although I guess there could there be a chance for a race trying to mount the volume onto the path before creating the user inside the container that creates the path? But that's probably a red herring for now.
    * Re: the fixed username in the current deployment, is that something that is set as a literal value at a single point in the k8s config (as the volume mount point, maybe?) and that same setting applies to everything it creates?
        * is there any way that name can be set from a variable? (I've deliberately avoided spending time learning k8s because I don't want to end up running deployments, so apologies for my ignorance...)

**Comment by mmh352** [comment]
1. Yes, that is correct.
2. Yes, that is the planned setup. The reason is that it is simply much easier maintenance-wise for me. In particular with the aim of having a single domain for access and integration into the single-sign-on. I might run a different setup if I were employed full-time to develop this solution, but part-time next to my actual duties minimal effort is the way to go. (Also it is a massive waste of resources. Each module would have base costs of 900 per presentation, which in the current case are shared evenly across all modules using the same JupyterHub).
3. The technical constraints hold for any JupyterHub on k8s installation. If SSU in their JupyterHub installation supported more than one container per JupyterHub installation, they would have to deal with the same problem.
   * The username is not necessarily needed in the mount point definition. However, the mount point location is defined once in the JupyterHub configuration and is the same for all containers run inside that JupyterHub.

Yes, all of these issues can be worked around, but each workaround is more complex and brittle than fixing the user-name and home-directory location.

**Reply by psychemedia** [reply]
Re: the costs/overheads of working with multiple JupyterHub servers, the JupyterHub server is quite lite so couldn't we run those locally and rely on cloud machines for user containers?

Re: pricing, when I looked at Jelastic, I got the feeling that they had an interesting pricing model and quite a lot of flexibility in terms of scaling vertically as well as horizontally. Plus a level of backend agnosticism which meant we wouldn't accidentally lock in to a particular cloud host (I can easily imagine hardwiring eg an Azure or AWS dependency in, then when they want to move everything in five years when we get a new head of IT in, there'll have to be a huge migration project effort...;-)

Re: resourcing: re: it being a lot for you to manage lots of hubs...  I'm not convinced. Eg if you have several modules with different presentation patterns, some of them updating containers, some of them not, students deferring etc, it could get messy; not like spin up a JupyterHub for a particular presentation and shut it down at the end etc.

Also, if there's container apps running across faculties, it may be that each faculty/school wants it's own setup. So planning for lots of deployments rather than one to rule them all would seem not unreasonable?

**Reply by mmh352** [reply]
The second point is the reason for running JupyterHub on k8s. You are not actually locking yourself into any vendor. Switching to another cloud provider only requires that they support running k8s and then migrating the students' persisted storage.

All of the other solutions are possible of course. However, they add a lot of complexity, but apart from being able to chuck any-old-container at the JupyterHub, I don't see them adding any value. If a scenario comes up that does not fit into the setup, then obviously, the setup needs to be adapted / reconsidered. However, I am not building and then maintaining huge, over-engineered solutions for requirements that do not exist.

What I mean is that that all of your suggestions are to avoid using the "ou-user" user and storing all user-persistent files in /home/ou-user/$CODE-$PRESENTATION. Implementing both in a container is not difficult and for me supporting the transition from an existing Jupyter image to an OU image is less effort than building and maintaining these massive solutions in which anything goes.

**Reply by psychemedia** [reply]
Sure, the simplifications all make sense.

I'm just flagging blockers as they occur to me for the ways that I intend to co-opt/appropriate/make use of ou-container-builder, eg in supporting containers I need to create for SSU or off-the-shelf JH setups with Jupyter defaults, or more generally elsewhere.

FWIW, I'm also happy to lobby SSU that they move to `ou-user` though there are issues w/ the `-` (if we are telling students to use eg the CLI to navigate to `ou-student` rather than `$HOME`, I've noticed students can misinterpret things with `-`, `_` etc symbols in them. Spaces appear, `-` can get omitted, etc.

**Reply by mmh352** [reply]
If you think that the username will be a maintenance/support problem, then that is something that can easily be changed. We could just change it to ``ou``. Short and concise. If you think this will help, then create an issue, as it is an easy fix.

**Reply by psychemedia** [reply]
Do you know what username IT have used? (I'm not sure if there demo server is still around?)

**Reply by mmh352** [reply]
No idea. I think they had a relatively off-the-shelf Jupyter container, so I would guess jovyan.

**Comment by psychemedia** [comment]
Just come across this wrt creating containers w/ username specified via the auth route:

> configure JupyterHub to extract the required LDAP attributes (username, UID), and pass them to the singleuser server by setting appropriate environment variables. If you start the singleuser server as root it will switch to that UID, which means it should be able to write to NFS as that user.

https://discourse.jupyter.org/t/starting-single-user-notebook-with-our-custom-ldap-docker-image/881/4

**Reply by psychemedia** [reply]
And also this for volume mounts: `The mountPath is templated, so you can use /home/{username}:`

**Reply by mmh352** [reply]
Interesting. I guess if you are fitting into an existing structure, where users have a directory on the NFS share, then that is really useful. In our scenario, where storage is allocated per-user by the cloud infrastructure, that is less relevant. In any case you still have the problem of multiple containers having to co-exist on the same filesystem. It would be the remote location, rather than the mount location that would have to be templated.
The nice thing is that if the remote location could be templated, then we could easily transition from the current solution to the templated solution, as from the container's point-of-view, it is only the location of the home directory that would change. As long as the paths aren't hard-coded, transitioning is possible.

**Comment by psychemedia** [comment]
Looking at the container on https://compute-home-test.teaching.room3b.eu I notice that for some reason I can't seem to upgrade the `nb_extension_empinken` extension myself properly. Extension files from `https://compute-home-test.teaching.room3b.eu/user/ajh59/nbextensions/empinken/index.js` don't seem to get updated? Unless there's an aggressive cache somewhere in the way?

**Reply by mmh352** [reply]
My guess would be that it is some kind of permissions issue.
